README.md
# gatsby-plugin-opensearch

Create an OpenSearch description for your Gatsby site.

_NOTE: This plugin only generates output when run in `production` mode! To test your OpenSearch description, run: `gatsby build && gatsby serve`_

## Install

`npm install gatsby-plugin-opensearch`

## How to Use

```javascript
// In your gatsby-config.js
siteMetadata: {
  siteUrl: `https://www.example.com`,
},
plugins: [{
  resolve: `gatsby-plugin-opensearch`,
  options: {
    shortName: 'Github',
    description: 'Search Github',
    searchTemplate: '/search?q={searchTerms}',
    searchForm: '/search',
  },
}]
```

Above is the minimal configuration required to get started. The
generated OpenSearch description will be included on all of your site's pages.

For more information about OpenSearch can be found [here](https://developer.mozilla.org/en-US/docs/Web/OpenSearch).

## Options

The `defaultOptions` [here](./src/defaults.js) can be overridden.

The options are as follows:

- `output` (string) The filepath and name. Defaults to `/opensearch.xml`.
- `shortName` (string) A short name for the search engine. It must be 16 or fewer characters of plain text, with no HTML or other markup.
- `description` (string) A brief description of the search engine. It must be 1024 or fewer characters of plain text, with no HTML or other markup.
- `searchTemplate` (string) The pathname along with any required query parameters. Parameters to be populated should be wrapped in curly brackets (e.g. `{searchTerms}`). More details of available parameters can be found [here](https://github.com/dewitt/opensearch/blob/master/opensearch-1-1-draft-6.md#opensearch-11-parameters). This string will be appended to the `siteUrl` to generate the OpenSearch description file. Any characters which are invalid in XML must be escaped (e.g. & -> `&amp;`)
- `searchForm` (string) The pathname for the site's search page. This allows Firefox users to visit the site directly.
- `createLinkInHead` (boolean) Whether to populate the `<head>` of your site with a link to the OpenSearch description file. Defaults to true.
- `image` (object) Allows configuration of an icon for the search engine (optional).
- `image.width` (number) The width of the icon
- `image.height` (number) The height of the icon
- `image.src` (string) The pathname for the icon, e.g. `/favicon.ico`. The site URL will be prefixed to the start


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [v1.4.0](https://github.com/garyforsterio/gatsby-plugin-opensearch/compare/v1.3.0...v1.4.0)

- feat: add option to specify image [`8b91ff0`](https://github.com/garyforsterio/gatsby-plugin-opensearch/commit/8b91ff0e0a950da90a3e91fcefb0ed929962913b)
- chore: remove crossorigin attribute [`a0d2e0e`](https://github.com/garyforsterio/gatsby-plugin-opensearch/commit/a0d2e0e3ec95787285ba7a09fa80db558bdd2ae8)
- feat: add title to link [`010560c`](https://github.com/garyforsterio/gatsby-plugin-opensearch/commit/010560cd34d87cbda9f716cc7139180b3c233e05)

#### [v1.3.0](https://github.com/garyforsterio/gatsby-plugin-opensearch/compare/v1.2.0...v1.3.0)

> 7 March 2022

- build(deps-dev): bump babel-preset-gatsby-package from 0.12.0 to 2.9.0 [`#9`](https://github.com/garyforsterio/gatsby-plugin-opensearch/pull/9)
- build(deps-dev): bump jest from 26.6.3 to 27.5.1 [`#6`](https://github.com/garyforsterio/gatsby-plugin-opensearch/pull/6)
- build(deps-dev): bump @babel/cli from 7.12.16 to 7.17.6 [`#5`](https://github.com/garyforsterio/gatsby-plugin-opensearch/pull/5)
- build(deps): bump @babel/runtime from 7.12.13 to 7.17.2 [`#4`](https://github.com/garyforsterio/gatsby-plugin-opensearch/pull/4)
- build(deps-dev): bump auto-changelog from 2.2.1 to 2.4.0 [`#7`](https://github.com/garyforsterio/gatsby-plugin-opensearch/pull/7)
- build(deps-dev): bump @babel/core from 7.12.16 to 7.17.5 [`#8`](https://github.com/garyforsterio/gatsby-plugin-opensearch/pull/8)
- feat: configure dependabot [`3d11954`](https://github.com/garyforsterio/gatsby-plugin-opensearch/commit/3d1195427842ed4db8c39107d359c61bcdc92514)
- feat: update gatbsy version [`3214690`](https://github.com/garyforsterio/gatsby-plugin-opensearch/commit/3214690a1577d721887f7dd2fc9092f8bb166963)

#### [v1.2.0](https://github.com/garyforsterio/gatsby-plugin-opensearch/compare/v1.1.0...v1.2.0)

> 5 April 2021

- feat: support gatsby v3 [`#1`](https://github.com/garyforsterio/gatsby-plugin-opensearch/pull/1)

#### [v1.1.0](https://github.com/garyforsterio/gatsby-plugin-opensearch/compare/v1.0.0...v1.1.0)

> 24 February 2021

- feat: add cross origin option [`5fdd8f6`](https://github.com/garyforsterio/gatsby-plugin-opensearch/commit/5fdd8f6ac2fb5c4bd164b4675734669a4c0dfece)

### [v1.0.0](https://github.com/garyforsterio/gatsby-plugin-opensearch/compare/v0.1.3...v1.0.0)

> 19 February 2021

- feat: add auto-changelog [`bbe7608`](https://github.com/garyforsterio/gatsby-plugin-opensearch/commit/bbe7608b188690b3e64caef2afc02929ab76a210)
- feat: add xml escaping [`870cc51`](https://github.com/garyforsterio/gatsby-plugin-opensearch/commit/870cc51c495ad532bef9595d8aed4ea5b2f2f025)

#### [v0.1.3](https://github.com/garyforsterio/gatsby-plugin-opensearch/compare/v0.1.2...v0.1.3)

> 19 February 2021

- fix: remove asset prefixer [`c10c099`](https://github.com/garyforsterio/gatsby-plugin-opensearch/commit/c10c09946b4468bc6c058eac886bc0917bd55217)

#### v0.1.2

> 18 February 2021

- feat: add jest [`2c0fbe5`](https://github.com/garyforsterio/gatsby-plugin-opensearch/commit/2c0fbe52e6f49ab4d23194077d6710adc21a136e)
- feat: setup repo [`091de1c`](https://github.com/garyforsterio/gatsby-plugin-opensearch/commit/091de1cd53882382690bd559004f270a8ac7881c)
- fix: remove unneccesary dependencies and add gatsby as peer [`ac0fc14`](https://github.com/garyforsterio/gatsby-plugin-opensearch/commit/ac0fc1483853376d745203a1dc444b7dbc0ca952)


src/createDescriptionFile.spec.js
import { createDescriptionFile } from './createDescriptionFile';
import { promises } from 'fs';

const fileName = 'test.xml';

const fields = {
  shortName: 'Shrt',
  description: 'Search engine',
  searchTemplate: '/test{searchParams}',
  searchForm: '/search',
  siteUrl: 'http://example.com',
};

const expectedXML = `<OpenSearchDescription xmlns="http://a9.com/-/spec/opensearch/1.1/" xmlns:moz="http://www.mozilla.org/2006/browser/search/">
<ShortName>Shrt</ShortName>
<Description>Search engine</Description>
<InputEncoding>UTF-8</InputEncoding>
<Url type="text/html" method="get" template="http://example.com/test{searchParams}"/>
<moz:SearchForm>http://example.com/search</moz:SearchForm>
</OpenSearchDescription>`;

describe('createDescriptionFile', () => {
  it('produces expected xml file', () => {
    const writeMock = jest.fn();
    jest.spyOn(promises, 'writeFile').mockImplementation(writeMock);
    createDescriptionFile(fileName, fields);
    expect(writeMock).toHaveBeenCalledTimes(1);
    expect(writeMock).toHaveBeenCalledWith(fileName, expectedXML);
  });
  it('populates image if specified', () => {
    const writeMock = jest.fn();
    jest.spyOn(promises, 'writeFile').mockImplementation(writeMock);
    createDescriptionFile(fileName, {
      ...fields,
      image: { width: 16, height: 16, src: '/favicon.ico' },
    });
    expect(writeMock).toHaveBeenCalledTimes(1);
    expect(writeMock).toHaveBeenCalledWith(
      fileName,
      expect.stringContaining(
        '<Image width="16" height="16" type="image/x-icon">http://example.com/favicon.ico</Image>'
      )
    );
  });
});


src/createDescriptionFile.js
import { promises as fs } from 'fs';
import xmlescape from 'xml-escape';

export const createDescriptionFile = (
  outputFile,
  { shortName, description, siteUrl, searchTemplate, searchForm, image }
) => {
  const components = [
    `<OpenSearchDescription xmlns="http://a9.com/-/spec/opensearch/1.1/" xmlns:moz="http://www.mozilla.org/2006/browser/search/">`,
    `<ShortName>${xmlescape(shortName)}</ShortName>`,
    `<Description>${xmlescape(description)}</Description>`,
    `<InputEncoding>UTF-8</InputEncoding>`,
    `<Url type="text/html" method="get" template="${xmlescape(
      siteUrl + searchTemplate
    )}"/>`,
    image
      ? `<Image width="${image.width}" height="${
          image.height
        }" type="image/x-icon">${xmlescape(siteUrl + image.src)}</Image>`
      : '',
    `<moz:SearchForm>${xmlescape(siteUrl + searchForm)}</moz:SearchForm>`,
    `</OpenSearchDescription>`,
  ].filter((elem) => !!elem);

  fs.writeFile(outputFile, components.join('\n'));
};


src/gatsby-node.js
import path, { join } from 'path';
import { createDescriptionFile } from './createDescriptionFile';
import { defaultOptions } from './config';

const publicPath = `./public`;

exports.onPostBuild = async ({ graphql }, pluginOptions) => {
  const { output, ...rest } = {
    ...defaultOptions,
    ...pluginOptions,
  };

  const { errors, data } = await graphql(`
    {
      site {
        siteMetadata {
          siteUrl
        }
      }
    }
  `);

  if (errors) {
    throw new Error(errors.join(`, `));
  }

  const siteUrl = data.site.siteMetadata.siteUrl;

  if (!siteUrl || siteUrl.trim().length == 0) {
    throw new Error(
      `SiteMetaData 'siteUrl' property is required and cannot be left empty.`
    );
  }

  const outputFile = path.join(publicPath, output);

  return createDescriptionFile(outputFile, { siteUrl, ...rest });
};

exports.pluginOptionsSchema = ({ Joi }) =>
  Joi.object().keys({
    output: Joi.string()
      .default(`/opensearch.xml`)
      .description(`The filepath and name`),
    createLinkInHead: Joi.boolean()
      .default(true)
      .description(
        `Whether to populate the \`<head>\` of your site with a link to the OpenSearch description.`
      ),
    shortName: Joi.string()
      .max(16)
      .required()
      .description(
        `A short name for the search engine. It must be 16 or fewer characters of plain text, with no HTML or other markup.`
      ),
    description: Joi.string()
      .max(1024)
      .required()
      .description(
        `A brief description of the search engine. It must be 1024 or fewer characters of plain text, with no HTML or other markup.`
      ),
    searchTemplate: Joi.string()
      .required()
      .description(
        'The pathname along with any required query parameters to be populated.'
      ),
    searchForm: Joi.string()
      .uri({ relativeOnly: true })
      .required()
      .description('The pathname of the search form.'),
    image: Joi.object().keys({
      width: Joi.number().required().description('The width of the icon'),
      height: Joi.number().required().description('The height of the icon'),
      src: Joi.string()
        .required()
        .description(
          'The pathname for the icon. The site URL will be prefixed to the start'
        ),
    }),
  });
